File: akit/src/main/java/org/littletonrobotics/junction/ReceiverThread.java
// Copyright (c) 2021-2025 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

package org.littletonrobotics.junction;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;

class ReceiverThread extends Thread {
  private final BlockingQueue<LogTable> queue;
  private List<LogDataReceiver> dataReceivers = new ArrayList<>();

  ReceiverThread(BlockingQueue<LogTable> queue) {
    super("AdvantageKit_LogReceiver");
    this.setDaemon(true);
    this.queue = queue;
  }

  void addDataReceiver(LogDataReceiver dataReceiver) {
    dataReceivers.add(dataReceiver);
  }

  public void run() {
    // Start data receivers
    for (int i = 0; i < dataReceivers.size(); i++) {
      dataReceivers.get(i).start();
    }

    try {
      while (true) {
        LogTable entry = queue.take(); // Wait for data

        // Send data to receivers
        for (int i = 0; i < dataReceivers.size(); i++) {
          dataReceivers.get(i).putTable(entry);
        }
      }
    } catch (InterruptedException exception) {
      // Empty queue
      while (!queue.isEmpty()) {
        LogTable entry = queue.poll();
        for (int i = 0; i < dataReceivers.size(); i++) {
          try {
            dataReceivers.get(i).putTable(entry);
          } catch (InterruptedException e) {
          }
        }
      }

      // End all data receivers
      for (int i = 0; i < dataReceivers.size(); i++) {
        dataReceivers.get(i).end();
      }
    }
  }
}

File: akit/src/main/native/cpp/ReceiverThread.cpp
// Copyright (c) 2021-2026 Littleton Robotics
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by a BSD
// license that can be found in the LICENSE file
// at the root directory of this project.

#include "akit/ReceiverThread.h"

using namespace akit;

ReceiverThread::ReceiverThread(
		moodycamel::BlockingReaderWriterCircularBuffer<LogTable> &queue) : queue { queue } {
}

void ReceiverThread::Start() {
	thread = std::make_unique < std::jthread > ([this](std::stop_token stopToken) { Run(stopToken); });
}

void ReceiverThread::End() {
	running = false;
	thread->join();
	thread.release();
}

void ReceiverThread::AddDataReceiver(
		std::unique_ptr<LogDataReceiver> receiver) {
	dataReceivers.emplace_back(std::move(receiver));
}

void ReceiverThread::Run(std::stop_token stopToken) {
	for (auto &receiver : dataReceivers)
		receiver->Start();

	std::optional<LogTable> entry;
	while (!stopToken.stop_requested()) {
		if (queue.wait_dequeue_timed(entry, std::chrono::milliseconds{100}))
			for (auto &receiver : dataReceivers)
				receiver->PutTable(*entry);
	}

	while (queue.try_dequeue(entry)) {
		for (auto &receiver : dataReceivers)
			receiver->PutTable(*entry);
	}
}
